text/0253-cdk-metadata-v2.md

```diff
@@ -0,0 +1,488 @@
+allow-list of packages (or package prefixes) we are going to report on. For
+example, the current list looks like this and can be found
+[here](https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/core/lib/private/runtime-info.ts):
+
+```ts
+const WHITELIST_SCOPES = ['@aws-cdk', '@aws-solutions-konstruk', '@aws-solutions-constructs', '@amzn'];
+const WHITELIST_PACKAGES = ['aws-rfdk'];
+```
+
+Having a PR opened against the CDK to update this list will be a clear
+interaction moment between an AWS team trying to onboard with this mechanism
+and us, and will prevent 3rd party authors from accidentally (or maliciously)
+onboarding themselves as well and sending us information that we don't want to
+be responsible for.
+
+## Metadata resource encoding
+
+Once we have the list of construct identifiers, we need to encode it into the
+metadata resource.
+
+The Metadata resource looks like this:
+
+```yaml
+Resources:
+  CDKMetadata:
+    Type: AWS::CDK::Metadata
+    Properties:
+      Modules: <v1 module information, deprecated by this RFC>
+      Analytics: <v2 construct information, introduced by this RFC>
+```
+
+### Construct list size
+
+The list of construct identifiers we have discovered may be sizeable. A rough
+estimate of the maximum number of constructs used in a stack will be about
+400 (maximum of 200 L1s with a corresponding L2 wrapper). We will therefore
+need to encode roughly 400 strings of the form
+`@aws-solutions-constructs/aws-lambda-sqs.LambdaToSqs@1.63.0`.
+
+The longest overall type in the current monocdk is
+`monocdk-experiment.aws_cognito.CfnUserPoolRiskConfigurationAttachment.CompromisedCredentialsRiskConfigurationTypeProperty`
+(121 characters), the longest construct class name is
+`monocdk-experiment.aws_ecs_patterns.ApplicationMultipleTargetGroupsFargateService`
+(81 characters).
+
+If an identifier is estimated to be a maximum of 100 bytes (which seems
+reasonable be plenty of room) the resulting data set will be at most ~40kB.
+
+We can encode the elements as a list straight in the template, but it will
+take up a lot of space in the template: vertical space while looking at it,
+as well as bytes that count toward the template size (maximum of 460kB).
+
+### Compression
+
+As an experiment, I've drawn 400 random class names from the complete list of class
+names currently in monocdk; that set of class names comes to about 23kB (uncompressed).
+
+Trying various methods of compression on it yields the following (keeping in mind
+that the result of binary compression will need to be base64-encoded to transmit via
+a CloudFormation template):
+
+| Description                | Raw size    | Base64ed size |
+|----------------------------|-------------|---------------|
+| Plaintext                  | 23k         | -             |
+| gzipped                    | 3.6k        | 4.8k          |
+| Zopfli                     | 3.4k        | 4.6k          |
+| bzip2                      | 3.4k        | 4.6k          |
```

Comment:
Curious if `zstd` is significantly different to those... Although this is probably just misplaced curiosity 🙃

Additionally - how's those configured? Default? There's a world of difference between `gzip -9` and `gzip -1` for example.

Reply:
It's all max compression